### ISSUE TITLE
[BUG] extern overrides giving compile errors

### ISSUE THREAD

**Issue by xposure** [beefytech/Beef#716]
https://github.com/xposure/Atma.Framework/commit/dbb2edfb0bd13b36c9bf70e155216f355f1d892f

If you open the workspace up at `/src`, the IDE reports compile errors for extern overrides, if you open workspace up at `/samples` the workspace builsd fine and loads the game. The only difference I can tell is that `src` doesn't have a launch/startup object.

**Comment by xposure** [comment]
Just updating that I created a new workspace called `Game` in its own folder and added an existing project to https://github.com/xposure/Atma.Framework/tree/master/src/Atma.Common 2731c79d4095dc54f6f698a151d4aad0692fab1b

**Comment by bfiete** [comment]
Working on a fix to check in tomorrow...

**Comment by xposure** [comment]
Ok!

I also noticed that if you close the IDE and reopen it the project goes red as if it couldn't be loaded. If you remove and re-add the project you get the compile errors again. Then closing and opening the IDE shows the project red again like its missing.

![image](https://user-images.githubusercontent.com/6292318/102028780-cdeb0200-3d79-11eb-9408-a9b48e05bf84.png)

Additionally, when it gets in this state, its not just extern, I can't use an extension class either.

![image](https://user-images.githubusercontent.com/6292318/102028774-c9264e00-3d79-11eb-960f-3466f9520426.png)

**Comment by xposure** [comment]
The IDE is now in a completely unusable state some how :/ I was working around it by just coding in the sample project until this error popped up. Whats really strange is that I had a decent amount of code in this file before it started crashing and even if I take the code back to bare bones and have 99% of it commented out, it still crashses.

Not sure if its related to this.

[gamecrash.zip](https://github.com/beefytech/Beef/files/5685902/gamecrash.zip)

**Comment by bfiete** [comment]
Is that crash from a version you built yourself?

**Comment by xposure** [comment]
I think it was built thursday or friday, I'm fairly sure its the latest commit though. But yea, I switched over to source builds now.

**Comment by xposure** [comment]
Just hopped back on this machine, its using the latest commit 9c948d37ffee0b993a6d798edb9e63ce47893c5b

**Comment by xposure** [comment]
Tried starting over and wrote this file, when I went to build the IDE crashed with [buildcrash.zip](https://github.com/beefytech/Beef/files/5686119/buildcrash.zip)

Relaunching did not trigger the error.

```
using Atma;
namespace NeonShooter
{
	class TurnGame : Scene
	{
		const int tiles = 5;
		const int tileSize = 32;
		const int gridSize = tiles * tileSize;

		public override void Initialize()
		{
			base.Initialize();
			Core.Graphics.ClearColor = .(51, 51, 51, 255);
			this.Camera.SetDesignResolution(gridSize, gridSize, .ShowAllPixelPerfect);
		}

		public override void FixedUpdate()
		{
			base.FixedUpdate();
		} protected internal override void Render()
		{
			Core.Draw.Rect(rect.FromDimensions(.Zero, .(gridSize)), .White);

			base.Render();
		}
	}
}
```

```

	public static class Program
	{
		static public int Main(String[] args)
		{
			//return Core.RunScene<NeonGame>("NeonShooter", 1280, 760);
			return Core.RunScene<TurnGame>("TurnGame", 1280, 760);
		}
	}
```

**Comment by bfiete** [comment]
I can't use crash dumps unless I have exact PDBs, DLLs, and EXEs from the crash. So no point in even sending crash reports from a custom build, really.

**Comment by xposure** [comment]
Gotcha, last update on this for the night. I sometimes see this in the build output.

```
LINK : warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library
```

**Comment by bfiete** [comment]
The fix is buried in a larger commit, here:
https://github.com/beefytech/Beef/commit/6d058ba59cef3b83644b545ddc0ddc65b46b81f4#diff-89d8614e0696e9ad93c75ca9b1187b89499392486209985c6882db645984a40b

The issue was a method ordering issue where both the extern declaration and override definition are both defined in extensions. The parsing order can change between workspaces (which is probably something that should be changed), and it worked if the extern was ordered first but it broke if the overload was ordered first.

I don't know what caused the additional issues like the red project. I need more info on that.

**Comment by bfiete** [comment]
Oh. And if you DO notice the "larger commit". It's a very large in-progress feature. Probably the biggest feature this year! Though it probably wont' actually be finished this year.

**Comment by xposure** [comment]
> Tried starting over and wrote this file, when I went to build the IDE crashed with buildcrash.zip
Here is a branch for repro https://github.com/xposure/Atma.Framework/tree/stb_sprintf

> I don't know what caused the additional issues like the red project. I need more info on that.
Working on a repro for this one.

**Comment by xposure** [comment]
> I don't know what caused the additional issues like the red project. I need more info on that.

This is reproduced in the `MissingGame` branch. If you load the workspace up at `Atma.Framework\UntitledGame\src` you'll notice the game is red, you can remove it and add it back from `Atma.Framework\UntitledGame\src\BeefProj.toml` once you close and reopen the IDE the issue will come back. It looks like the workspace is saving the path to the project as `UntitledGame\src\BeefProj.toml\BeefProj.toml`

**Comment by bfiete** [comment]
Okay, so for this gamecrash in the stb_sprintf branch -- what do I do to reproduce this bug?

**Comment by xposure** [comment]
The IDE crashes on open for me (using latest commit but also prior commit from last week).

**Comment by xposure** [comment]
Updated the opening comment to contain a list of what this is tracking.

I mentioned there was an issue with an error "unable to find root type". This is reproduced in ExtensionBug branch.

**Comment by xposure** [comment]
[stb.zip](https://github.com/beefytech/Beef/files/5689203/stb.zip)

Using latest nightly

**Comment by bfiete** [comment]
Yeah, a reproducible steps would be nice for the 'root type' error

**Comment by xposure** [comment]
I just pushed a commit to the `ExtesionBug` branch, after resolving all the issues with root type and dependencies, its reporting a linker issue with tuples.

**Comment by bfiete** [comment]
stb crash fixed at https://github.com/beefytech/Beef/commit/f28229aef4b964304ad51dfa0cf0dfbe7f473693
'red project' fixed at https://github.com/beefytech/Beef/commit/9560a8de88b61e375b6b72ba545b28c78cbf7066

**Comment by bfiete** [comment]
Tuple issue fixed with a clean, I'm guessing? I'd love a repro for that bug I'm guessing that will be pretty tricky.

**Comment by xposure** [comment]
Clean doesn't fix it.

**Comment by bfiete** [comment]
Hm. I just fixed the one dependency in HelloWorld in that branch and everything linked fine

**Comment by bfiete** [comment]
Oh. ExtensionBug. I'll try that.

**Comment by xposure** [comment]
Steps to reproduce `unable to find root type`

1. Checkout master branch
2. Open workspace samples
3. New Project -> `Atma.Framework\src\Atma.Entities` names `Atma.Entities` type `Library`
4. Right click `Atma.Common` open containing folder and cut all the files+dir inside `src/Entities`
5. Paste the files in `Atma.Framework\src\Atma.Entities\src`
6. Go back to the IDE and notice the error output window. These errors will go away if you reload the IDE.

**Comment by bfiete** [comment]
Tuple link bug fixed at https://github.com/beefytech/Beef/commit/cdc030ab30416db7d208f1ec689df29f308ed5f8
